plugin/restore/mysql/mysql.go

```diff
@@ -457,87 +480,183 @@ func getPITROldDatabaseName(database string, suffixTs int64) string {
+		}
+	}
+	binlogFilesLocalValidSorted, err := sortBinlogFiles(binlogFilesLocalValid)
 	if err != nil {
-		return err
+		return nil, nil, err
+	}
+
+	return downloadFileList, binlogFilesLocalValidSorted, nil
+}
+
+func convertToBinlogFiles(binlogFileInfoList []fs.FileInfo) ([]BinlogFile, error) {
+	var binlogFileList []BinlogFile
+	for _, fileInfo := range binlogFileInfoList {
+		binlogFile, err := newBinlogFile(fileInfo.Name(), fileInfo.Size())
+		if err != nil {
+			return nil, err
+		}
+		binlogFileList = append(binlogFileList, binlogFile)
 	}
+	return binlogFileList, nil
+}
 
-	binlogFilesOnServer, err := r.getBinlogFilesMetaOnServer(ctx)
+// FetchBinlogFilesToTargetTs downloads the locally missing binlog files from the remote instance to `binlogDir`.
+// After downloading a binlog file, we check it's first event timestamp. If the timestamp is larger than
+// targetTs, we stop the following downloads and return eagerly, because now the local binlog files contain
+// all the binlog events we need for this recovery task.
+func (r *Restore) FetchBinlogFilesToTargetTs(ctx context.Context, targetTs int64) error {
+	// Read the local binlog files.
+	binlogFilesInfoLocal, err := ioutil.ReadDir(r.binlogDir)
 	if err != nil {
 		return err
 	}
-
-	latestBinlogFileOnServer, err := r.getLatestBinlogFileMeta(ctx)
+	binlogFilesLocal, err := convertToBinlogFiles(binlogFilesInfoLocal)
 	if err != nil {
 		return err
 	}
 
-	// build a local file size map from file name to size
-	localFileMap := make(map[string]int64)
-
-	for _, localFile := range binlogFilesLocal {
-		localFileMap[localFile.Name()] = localFile.Size()
+	// Read binlog files list on server.
+	// We compare it with the local binlog files list to generate a download list.
+	// Also compare file sizes to validate local binlog files, and re-download invalid ones.
+	binlogFilesOnServerSorted, err := r.GetSortedBinlogFilesMetaOnServer(ctx)
+	if err != nil {
+		return err
+	}
+	if len(binlogFilesOnServerSorted) == 0 {
+		// No binlog files on server, so there's nothing to download.
+		return nil
 	}
+	latestBinlogFileOnServer := binlogFilesOnServerSorted[len(binlogFilesOnServerSorted)-1]
+	log.Debug("Got sorted binlog file list on server", zap.Array("list", ZapBinlogFiles(binlogFilesOnServerSorted)))
 
-	todo := make(map[string]bool)
+	downloadFileList, binlogFilesLocalValidSorted, err := getBinlogDownloadListAndLocalValidFiles(r.binlogDir, binlogFilesLocal, binlogFilesOnServerSorted)
+	if err != nil {
+		return err
+	}
 
-	for _, serverBinlog := range binlogFilesOnServer {
-		// We don't download the latest binlog in SyncArchivedBinlogFiles()
-		if serverBinlog.Name == latestBinlogFileOnServer.Name {
-			continue
+	// Find if some local binlog file's first binlog event ts is already larger than targetTs.
+	// If so, we do not need to download the following binlog files on server.
+	var foundLocalFileLTTargetTs bool
+	var localFileExceedsTargetTs BinlogFile
```

Comment:
we only need the seq instead of the whole BinlogFile

Reply:
Done.